hw2_1/HW_1.py

```diff
@@ -0,0 +1,713 @@
+# 
+# Вся сложность задания в том, чтобы правильно организовать код. Если у вас есть повторяющийся код в сестринских классах или родительском и дочернем, значит вы что-то делаете не так.
+# 
+# 
+# Маленькое замечание: По-хорошему, между классом `BiologicalSequence` и классами `NucleicAcidSequence` и `AminoAcidSequence`, ещё должен быть класс-прослойка, частично реализующий интерфейс `BiologicalSequence`, но его писать не обязательно, так как задание и так довольно большое (правда из-за этого у вас неминуемо возникнет повторяющийся код в классах `NucleicAcidSequence` и `AminoAcidSequence`)
+
+# In[41]:
+
+
+from abc import ABC, abstractmethod
+
+class BiologicalSequence(ABC):
+    
+    @abstractmethod
+    def __len__(self):
+        pass
+
+    @abstractmethod
+    def __getitem__(self, slc):
+        pass
+    
+    @abstractmethod
+    def __str__(self):
+        pass
+    
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+    
+    
+class SequenceIncorrectionError(Warning):  
+    
+    def __str__(cls):
+        return 'Your sequence does not match with alphabet. See standart alphabet of class using attribute .alphabet'
+
+
+# In[42]:
+
+
+class ClassicalPolymer(BiologicalSequence):
+    
+    def __init__(self, sequence):
+        self.sequence = sequence
+        self.standard_alphabet = {'DNA' : {'t', 'g', 'c', 'a'},
+                                  'RNA' : {'u', 'g', 'c', 'a'},
+                                  'Protein' : {'a', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'k', 'l', 'm', 'n','p', 'q', 'r', 's', 't', 'v', 'w', 'y'}}
+    
+    def __len__(self):
+        return len(self.sequence)
+
+    def __getitem__(self, slc):
+        return type(self)(self.sequence[slc])
+    
+    def __str__(self):
+        return str(self.sequence)
+    
+    def __repr__(self):
+        return 'Sequence: ' + str(self.sequence)
+    
+    def check_alphabet(self):
+        if isinstance(self, DNASequence):
+            self.alphabet = self.standard_alphabet['DNA']
+        elif isinstance(self, RNASequence):
+            self.alphabet = self.standard_alphabet['RNA']
+        elif isinstance(self, AminoAcidSequence):
+            self.alphabet = self.standard_alphabet['Protein']
+        else:
+            raise BaseException('Nonstandard input')   # in case of carbohydrates in future
+            
+        if set(self.sequence.lower()) <= self.alphabet:
+            return True
+        else:
+            raise SequenceIncorrectionError
+
+
+# In[43]:
+
+
+class NucleicAcidSequence(ClassicalPolymer):
+    
+    def complement(self):
+        if isinstance(self, DNASequence) or isinstance(self, RNASequence): 
+            table = str.maketrans('aAgGcCtTuU', 'tTcCgGaAaA')
+            return type(self)(self.sequence.translate(table))
+        else:
+            raise NotImplementedError('Only for DNA and RNA subclasses') 
```

Comment:
Тут всё аналогично `check_alphabet`. Удобно каждому из дочерних классов сделать атрибут с правилами комплементирования, а здесь (в родительском, его использовать), тогда условия не нужны
```suggestion
        if not hasattr(self, "table"):
            raise NotImplementedError(f'Class {type(self) must implement "table" attribute to make complement sequence}') 
        return type(self)(self.sequence.translate(self.table))            
```